### PR TITLE
encoding/decoding base64

### PR DESCRIPTION
**Pull request by rap2hpoutre**
This recipe shows how to encode and decode base64 (https://github.com/brson/rust-cookbook/issues/116). This example is a bit boring, I can improve it if required (but I have inspiration right now). Feel free to criticize, I'm still learning.

### CODE REVIEW

File: src/encoding.md [hunk 1]
```diff
@@ -225,6 +226,49 @@ to be percent-encoded. The choice of this set depends on context. For example,
...
 collected into a `String`.
 
+[ex-base64]: #ex-base64
+<a name="ex-base64"></a>
+## Encode and decode base64
+
+[![base64-badge]][base64] [![cat-encoding-badge]][cat-encoding]
+
+The [`base64`] crate provides `encode` and `decode` functions.
+The example below encode a byte slice to a `base64` encoded `String`
```
Comment (budziq) [inline src/encoding.md]: I would make this whole description more brief. Something like 
"Byte slice is incoded into `base64` String with help of `encode` and subsequently decoded with `decode`."

File: src/encoding.md [hunk 2]
```diff
@@ -225,6 +226,49 @@ to be percent-encoded. The choice of this set depends on context. For example,
...
+then decode it. The results are printed to the standard output.
+
+[`base64`]: https://docs.serde.rs/base64/
+
+```rust
+#[macro_use]
+extern crate error_chain;
+
```
Comment (budziq) [inline src/encoding.md]: please remove the whitespace

File: src/encoding.md [hunk 3]
```diff
@@ -225,6 +226,49 @@ to be percent-encoded. The choice of this set depends on context. For example,
...
+error_chain! {
+    foreign_links {
+        Base64(base64::DecodeError);
+        Utf8Error(str::Utf8Error);
+    }
+}
+
+fn run() -> Result<()> {
+    let hello = b"hello rustaceans";
```
Comment (budziq) [inline src/encoding.md]: no need for intermediate variable

File: src/encoding.md [hunk 4]
```diff
@@ -225,6 +226,49 @@ to be percent-encoded. The choice of this set depends on context. For example,
...
+    let encoded = encode(hello);
+    let decoded = &decode(&encoded)?[..];
```
Comment (budziq) [inline src/encoding.md]: I would suggest 
let decoded = decode(&encoded)?;
println!("back to origin: {}", str::from_utf8(&decoded)?);

**Comment by budziq** [comment]
Great work!. @rap2hpoutre  Also please try to git rebase -i these three commits to one commit. As these commits make no sens separately.

**Comment by rap2hpoutre** [comment]
:/ I broke everything.

**Comment by budziq** [comment]
@rap2hpoutre please do not giveup! You are almost there. Your biggest problem is using the github web-edit which is abysmal for complex changes and rebases.

Try to use command line git client. Do not be afraid to ask for help.

**Comment by rap2hpoutre** [comment]
@budziq Thanks (again)! I tried to resolve conlict+add new commit+rebase and I messed up everything. Now IMO my branch is a can of worms and the history is a mess. Do you think I should create a new branch+pull request (and do everything with command line and my text editor, +commit fix with amend) OR should I try to fix this branch history and reopen this PR? To be fair, I prefer the first option! I could work on it tomorow afternoon (utc+2). Anyway I'm still happy to contribute to various Rust (projects including this one) I'm learning **a lot** of things (including git!).

**Comment by budziq** [comment]
If you are not comfortable with some git workflows yet I would also suggest new branch and PR with git cli and editor.
In general I would suggest avoiding web-edit for any non-trivial changes (in this way you will be able to make a single commit for all modified files which is a big win when rebasing)

**Comment by rap2hpoutre** [comment]
@budziq I created a new PR here: https://github.com/brson/rust-cookbook/pull/135

> If you are not comfortable with some git workflows 

You are right, I'm not!

>  I would suggest avoiding web-edit for any non-trivial changes

:+1:

**Pull request closed without merging**